Top-Level Spec (#173)
ucan-wg/spec: README.md

### Certificate Capabilities

UCANs work more like [movie tickets][caps as keys] or a festival pass. No one needs to check your ID; who you are is irrelevant. For example, if you have a ticket issued by the theater to see Citizen Kane, you are admitted to Theater 3. If you cannot attend an event, you can hand this ticket to a friend who wants to see the film instead, and there is no coordination required with the theater ahead of time. However, if the theater needs to cancel tickets for some reason, they need a way of uniquely identifying them and sharing this information between them.

### Object Capabilities

Object capability ("ocap") systems use a combination of references, encapsulated state, and proxy forwarding. As the name implies, this is fairly close to object-oriented or actor-based systems. Object capabilities are [robust][Robust Composition], flexible, and expressive.

To achieve these properties, object capabilities have two requirements: [fail-safe], and locality preservation. The emphasis on consistency rules out partition tolerance[^pcec].

## Security Considerations

> Each UCAN includes an assertions of what it is allowed to do. "Proofs" are positive evidence (elsewhere called "witnesses") of the possession of rights. They are cryptographically verifiable chains showing that the UCAN issuer either claims to directly own a resource, or that it was delegated to them by some claimed owner. In the most common case, the root owner's ID is the only globally unique identity for the resource.

Reviewer: ```suggestion
Each UCAN includes assertions of what it is allowed to do. "Proofs" are positive evidence (elsewhere called "witnesses") of the possession of rights. They are cryptographically verifiable chains showing that the UCAN issuer either claims to directly own a resource, or that it was delegated to them by some claimed owner. In the most common case, the root owner's ID is the only globally unique identity for the resource.
```
"an" looks unneeded here.

Note that a structurally and cryptographically valid UCAN chain can be semantically invalid. The executor MUST verify the ownership of any external resources at execution time. While not possible for all use cases (e.g. replicated state machines and eventually consistent data), having the Executor be the resource itself is RECOMMENDED.

While certificate chains go a long way toward improving security, they do not provide [confinement] on their own. The principle of least authority SHOULD be used when delegating a UCAN: minimizing the amount of time that a UCAN is valid for and reducing authority to the bare minimum required for the delegate to complete their task. This delegate should be trusted as little as is practical since they can further sub-delegate their authority to others without alerting their delegator. UCANs do not offer confinement (as that would require all processes to be online), so it is impossible to guarantee knowledge of all of the sub-delegations that exist. The ability to revoke some or all downstream UCANs exists as a last resort.

## 1.4 Inversion of Control

> [Inversion of control] is achieved due to two properties: self-certifying delegation and reference passing. There is no Authorization Server (AS) that sits between requestors and resources. In traditional terms, the owner of a UCAN resource is the resource server (RS) directly.

Author: I can add words here to say that the user is the RS... it's mainly for people coming from OAuth.

FWIW Fission used to do it the way you described, and we've come to reguard that as a mistake 😅

A large portion of personal information now also moves through connected systems. As a result, data privacy is a prominent theme when considering the design of modern applications, to the point of being legislated in parts of the world. 

Ahead-of-time coordination is often a barrier to development in many projects. Flexibility to define specialized authorization semantics for resources and the ability to integrate with external systems trustlessly are essential as the number of autonomous, specialized, and coordinated applications increases.

Many high-value applications run in hostile environments. In recognition of this, many vendors now include public key functionality, such as [non-extractable keys in browsers][browser api crypto key], [certificate systems for external keys][fido], [platform keys][passkey], and [secure hardware enclaves] in widespread consumer devices.

Two related models that work exceptionally well in the above context are Simple Public Key Infrastructure ([SPKI][spki rfc]) and object capabilities ([OCAP]). Since offline operation and self-verifiability are two requirements, UCAN adopts a [certificate capability model] related to [SPKI].

> ## Intuition for Auth System Differences 

Reviewer 2: ```suggestion
## Intuition for Auth System Differences
```

### Certificate Capabilities

UCANs work more like [movie tickets][caps as keys] or a festival pass. No one needs to check your ID; who you are is irrelevant. For example, if you have a ticket issued by the theater to see Citizen Kane, you are admitted to Theater 3. If you cannot attend an event, you can hand this ticket to a friend who wants to see the film instead, and there is no coordination required with the theater ahead of time. However, if the theater needs to cancel tickets for some reason, they need a way of uniquely identifying them and sharing this information between them.

### Object Capabilities

Object capability ("ocap") systems use a combination of references, encapsulated state, and proxy forwarding. As the name implies, this is fairly close to object-oriented or actor-based systems. Object capabilities are [robust][Robust Composition], flexible, and expressive.

To achieve these properties, object capabilities have two requirements: [fail-safe], and locality preservation. The emphasis on consistency rules out partition tolerance[^pcec].

## Security Considerations

> Each UCAN includes an assertions of what it is allowed to do. "Proofs" are positive evidence (elsewhere called "witnesses") of the possession of rights. They are cryptographically verifiable chains showing that the UCAN issuer either claims to directly own a resource, or that it was delegated to them by some claimed owner. In the most common case, the root owner's ID is the only globally unique identity for the resource.

Reviewer 2: As [gozala mentioned a while ago](https://github.com/ucan-wg/spec/pull/173#pullrequestreview-1683120805), I think the concept of "proof chain" can be a bit of a weaselword for a first-time reader.  I would recommend adding a ### Proof Chains section defining what you do and don't mean by the term, immediately before ## Security Considerations and after ### Object Capabilities, since the ## Intuition section seems like the right place to set up novel primitives before returning to the macro-view in ## Security Considerations.

> Root capability issuers function as verifiable, distributed roots of trust. The delegation chain is by definition a provenance log. Private keys themselves SHOULD NOT move from one context to another. Keeping keys unique to each physical device and unique per use case is RECOMMENDED to reduce opportunity for keys to leak, and limit blast radius in the case of compromises. "Sharing authority without sharing keys" is provided by capabilities, so there is no reason to share keys directly.

Reviewer 2: The first sentence has always struck me as confusing, and I was afraid to say anything for fear of sounding dumb until now.  Is the root _capability_ itself the root of trust, or is the ID by which its issuer is known the root of trust? either way it feels like you're saying that a root capability should be trusted anywhere its issuer's ID is trusted, and can be verified/refreshed/checked-in-on...

| [Invocation] | Exercise authority that has been delegated through one or more delegates | REQUIRED          |
| [Promise]    | Await the result of an Invocation inside another Invocation              | RECOMMENDED       |
| [Revocation] | Undo a delegation, breaking a delegation chain for malicious users       | RECOMMENDED       |

``` mermaid
flowchart TD
    rev[Revocation]
    inv[Invocation]
    del[Delegation]
    prm[Promise]

    prm -->|awaits| inv
    rev -->|a kind of| inv -->|proven by| del
    rev -->|invalidates| del

    click del href "https://github.com/ucan-wg/delegation" "UCAN Delegation Spec"
    click inv href "https://github.com/ucan-wg/invocation" "UCAN Invocation Spec"
    click rev href "https://github.com/ucan-wg/revocation" "UCAN Revocation Spec"
```

## 2.1 Time

It is often useful to talk about a UCAN in the context of some action. For example, a UCAN delegation may be valid when it was created, but expired when invoked.

``` mermaid
sequenceDiagram
    Alice -->> Bob: Delegate
    Bob ->> Bob: Validate
    Bob -->> Carol: Delegate
    Carol ->> Carol: Validate
    Carol ->> Alice: Invoke
    Alice ->> Alice: Validate

>     Alice ->> Alice: Execute

Author: You should be able to check the UCAN chain at each hop, no?

## 3.1 Roles

There are several roles that an agent MAY assume:

| Name      | Description                                                                                      | 
|-----------|--------------------------------------------------------------------------------------------------|
| Agent     | The general class of entities and principals that interact with a UCAN                           |
| Audience  | The Principal delegated to in the current UCAN. Listed in the `aud` field                        |
| Executor  | The Agent that actually performs the action described in an invocation                           |
| Invoker   | A Principal that requests that an Executor perform some action that uses the Invoker's authority |
| Issuer    | The Principal of the current UCAN. Listed in the `iss` field                                     |
| Owner     | A Subject that controls some external resource                                                    |
| Principal | An agent identified by DID (listed in a UCAN's `iss` or `aud` field)                             |
| Revoker   | The Issuer listed in a proof chain that revokes a UCAN                                           |

> | Subject   | The Principal who's authority is delegated or invoked                                            |

Author: Subject is a standard field in many auth tokens

Unless explicitly stated, the Resource of a UCAN MUST be the Subject.

## 3.4 Command

Commands are concrete messages ("verbs") that MUST be unambiguously interpretable by the Subject of a UCAN. Commands are REQUIRED in invocations. Some examples include `msg/send`, `crud/read`, and `ucan/revoke`.

Much like other message-passing systems, the specific resource MUST define the behavior for a particular message. For instance, `crud/update` MAY be used to destructively update a database row, or append to a append-only log. Specific messages MAY be created at will; the only restriction is that the Executor understand how to interpret that message in the context of a specific resource.

While arbitrary semantics MAY be described, they MUST apply to the target resource. For instance, it does not make sense to apply `msg/send` to a typical file system. 

Commands MUST NOT be case-sensitive. There MUST be at least one path segment as a namespace. For example, `http/put` and `db/put` MUST be treated as unique from each other.

### 3.4.1 Ability

> Abilities abstract over [Command]s to allow for extension of UCAN delegations.

Author: Yeah this definitely needs to be tightened up. The difference is that a Command is something that can be called in an Invocation. You can't really call arguments on `http/*`, and adding any args to `http/*` make it much more difficult to delegate portion (`http/* -> http/put`) because it's really difficult to figure out how one would check the delegation.

I've been thinking about maybe just dropping the distinction and saying "too bad, you need to trace the `http/put` all the way to the root, you can't use `http/*` to prove it" but that seems like such a shame.

> Abilities MAY be organized in a hierarchy. A typical example is a superuser capability ("anything") on a file system. Another is read vs write access, such that in an HTTP context, `WRITE` implies `PUT`, `PATCH`, `DELETE`, and so on. Organizing abilities this way allows for adding more options over time in a backward-compatible manner, avoiding the need to reissue UCANs with new resource semantics.

Author: > I would highly recommend organizing hierarchies through name-spacing and direct reader to doin the same.

We've considered this, but it breaks when you want to be able to update these later. As described, you have to impose these fully up front syntactically. Validation already has to understand the proxying pattern (`ucan/*`) to connect disparate chains, so validtaion is not fully syntactically driven

A common use for the caveats field is to define the resource that sits behind the Subject.

## 3.6 Capability

A capability is the association of an ability to a subject: `subject x ability x caveats`.

The resource and ability fields are REQUIRED. Any non-normative extensions are OPTIONAL.

For example, a capability may used to represent the ability to send email from a certain address on Fridays:

| Field   | Example                                                    |
|---------|------------------------------------------------------------|
| Subject | `did:key:z6MkhaXgBZDvotDkL5257faiztiGiC2QtKLGpbnnEGta2doK` |
| Ability | `msg/send`                                                 |
| Caveat  | `{sender: "mailto:alice@example.com", "day": "friday"}`    |

> ## 3.7 Authority

Author: Authority is the noun — the act of "delegating authority" is "authorization"

1. Request protocol
2. Response protocol
3. Collections format

Note that if an instance cannot dereference a CID at runtime, the UCAN MUST fail validation. This is consistent with the [constructive semantics] of UCAN.

## 4.1 Content Identifiers

A UCAN token SHOULD be referenced as a [base32][multibase] [CIDv1]. [SHA2-256] is the RECOMMENDED hash algorithm. The [DAG-CBOR] codec MUST be supported, and [DAG-JSON] support is RECOMMENDED. 

The resolution of these addresses is left to the implementation and end-user, and MAY (non-exclusively) include the following: local store, a distributed hash table (DHT), gossip network, or RESTful service.

# 5. Implementation Recommendations

## 5.1 Delegation Store

A validator MAY keep a local store of UCANs that it has received. UCANs are immutable but also time-bound so that this store MAY evict expired or revoked UCANs.

This store SHOULD be indexed by CID (content addressing). Multiple indices built on top of this store MAY be used to improve capability search or selection performance.

## 5.2 Memoized Validation

> Aside from revocation, capability validation is idempotent. Marking a CID (or capability index inside that CID) as valid acts as memoization, obviating the need to check the entire structure on every validation. This extends to distinct UCANs that share a proof: if the proof was previously reviewed and is not revoked, it is RECOMMENDED to consider it valid immediately.

Author: > UCAN may include multiple capabilities 

Not anymore... they're single resource

> Revocation is irreversible. Suppose the validator learns of revocation by UCAN CID or issuer DID. In that case, the UCAN and all of its derivatives in such a cache MUST be marked as invalid, and all validations immediately fail without needing to walk the entire structure.

Author: Yeah that's true. This is from the old revocation format. Fixed.

## 5.3 Replay Attack Prevention

> Replay attack prevention is REQUIRED. Every UCAN token MUST have a unique CID. Some simple strategies for implementing uniqueness tracking include maintaining a set of previously seen CIDs, or requiring that nonces be monotonically increasing per principal. This MAY be the same structure as a validated UCAN memoization table (if one is implemented).

Author: >  Recommendation seems to be never issue the same UCAN which seems odd

Yeah, you should never use the same Invocation twice. They MUST be unique.

If they're not unique, then a PITM can intercept and replay the token. As a simple example for intuition: if your invocation says "increment counter" they can just run it thousand times even though the one submitting it doens't control the issuer's DID.

## 5.4 Beyond Single System Image

> As we continue to increase the number of globally connected devices, we must embrace a design that considers every single member in the system as the primary site for the data that it is generates. It is completely impractical that we can look at a single, or a small number, of globally distributed data centers as the primary site for all global information that we desire to perform computations with.
>
> —[Meiklejohn], [A Certain Tendency Of The Database Community]

Unlike many authorization systems where a service controls access to resources in their care, location-independent, offline, and leaderless resources require control to live with the user. Therefore, the same data MAY be used across many applications, data stores, and users. Since they don't have a single location, applying UCAN to [RSM]s and [CRDT]s MAY be modelled by lifting the requirement that the Executor be the Subject.

Ultimately this comes down to a question of push vs pull. In push, the subject MUST be the specific site being pushed to ("I command you to apply the following updates to your state").

> Pull is the broad class of situations where an Invoker doesn't require that a particular replica apply its state. Applying a change to a local CRDT replica and maintaining a UCAN invocation log is a valid update to "the CRDT": a version of the CRDT Subject exists locally even if the Subject's private key is not present. Gossiping these changes among agents allows each to apply changes that it becomes aware of. Thanks to the invocation log (or equivalent integrated directly into the CRDT), provenance of authority is made transparent.

Author: It's the difference between a "single system image" (AKA "location") versus a replicated agent. Take a WNFS or an Automerge document for example: there are two things that you may want to express:

1. I can write into this data staructure in general (even locally)
2. I'm allowed to update some specific replica (location)

(1) isn't in reference to any particular replica. I can make the update directly on my node, and attach the proofs for anyone to inspect if they **pull** the CRDT.

In (2), you **push** to a certain replica. You're consuming storage on a particular node, and works much more like a traditional HTTP PUT or POST request.

``` mermaid
sequenceDiagram
    participant CRDT as Initial Grow-Only Set (CRDT)

    actor Alice
    actor Bob
    actor Carol
    
    autonumber

    Note over CRDT, Bob: Setup
    CRDT -->> Alice: delegate(CRDT_ID, merge)
    CRDT -->> Bob: delegate(CRDT_ID, merge)
    
    Note over Bob, Carol: Bob Invites Carol
    Bob -->> Carol: delegate(CRDT_ID, merge)

    Note over Alice, Carol: Direct P2P Gossip
    Carol ->> Bob: invoke(CRDT_ID, merge, {"Carrot"}, proof: [➋,➍])
    Alice ->> Carol: invoke(CRDT_ID, merge, {"Apple"}}, proof: [➊])

>     Bob ->> Alice: invoke(CRDT_ID, merge, {"Banana", "Carrot"}, proof: [➋])

Author: @Gozala Right, I'm trying to show a different use case. You're treating it as a physical location, as opposed to a replicated agent

└─────────────┘   └─────────────┘   └─────────────┘
```

> This additionally allows UCAN to model auth for [eventually consistent and replicated state][overcoming SSI].

Reviewer 2: ```suggestion
This additionally allows UCAN to model authorization for [eventually consistent and replicated state], thus [overcoming SSI] constraints.
```

> # Roles

Reviewer 2: ```suggestion
## Roles
```
I think single-hash renames the file in some forms of markdown? in any case my markdown linter plugin discourages using them within a document

There are several roles that an agent MAY assume:

> | Name      | Description                                                                                      | 

Reviewer 2: ```suggestion
| Name      | Description                                                                                      |
```
linting like crazy over here

|-----------|--------------------------------------------------------------------------------------------------|
| Agent     | The general class of entities and principals that interact with a UCAN                           |
| Audience  | The Principal delegated to in the current UCAN. Listed in the `aud` field                        |
| Executor  | The Agent that actually performs the action described in an invocation                           |

> | Invoker   | A Principal that requests an Executor perform some action that uses the Invoker's authority |

Reviewer 2: ```suggestion
| Invoker   | A Principal that requests an Executor perform some action that uses the Invoker's authority      |
```

| Issuer    | The Principal of the current UCAN. Listed in the `iss` field                                     |

> | Owner     | A Subject that controls some external resource                                                    |

Reviewer 2: ```suggestion
| Owner     | A Subject that controls some external resource                                                   |
```

### Resource

A resource is some data or process that can be uniquely identified by a [URI]. It can be anything from a row in a database, a user account, storage quota, email address, etc. Resource MAY be as coarse or fine grained as desired. Finer-grained is RECOMMENDED where possible, as it is easier to model the principle of least authority ([PoLA]).

A resource describes the noun of a capability. The resource pointer MUST be provided in [URI] format. Arbitrary and custom URIs MAY be used, provided that the intended recipient can decode the URI. The URI is merely a unique identifier to describe the pointer to — and within — a resource.

Having a unique agent represent a resource (and act as its manager) is RECOMMENDED. However, to help traditional ACL-based systems transition to certificate capabilities, an agent MAY manage multiple resources, and [act as the registrant in the ACL system][wrapping existing systems].

> Unless explicitly stated, the Resource of a UCAN MUST be the Subject.

Reviewer 2: I don't understand where and how any other state of affairs could be explicitly stated? Do you mean 

```suggestion
Unless all expected recipients of a UCAN will understand how the Subject of a UCAN relates to the Resource being authorized because this relationship is explicitly clarified out-of-band, it can be assumed by all parties that a UCAN's subject is coterminous with the Resource authorized by that UCAN.
```
?  I assume not, but if you asked _me_ to make that sentence more verbose, this is what I would come up with, which feels like a very wrong guess.

## Issuer & Audience

> The Issuer (`iss`) and Audience (`aud`) can be conceptualized as the sender and receiver (respectively) of a postal letter. Every UCAN MUST be signed with the private key associated with the DID in the `iss` field.

Reviewer 2: ```suggestion
The Issuer (`iss`) and Audience (`aud`) can be conceptualized as the sender and receiver (respectively) of a postal letter. Every UCAN MUST be signed with the private key associated with the DID in the `iss` field. 

Note that the UCAN system assumes a 1:1 relationship between DID URLs and public keys, and all examples are given using `did:key`, where bare DIDs are resolvable directly and deterministically to single keys. For DID methods that support multiple keys for a given DID, verification-specific [DID URLs using fragment identifiers](https://www.w3.org/TR/did-core/#fragment) MUST be used for `iss` and `aud` values to preserve interoperability; non-key-based `verificationMethod`s MUST NOT be used, for the same reason.
```
I think I promised this as a PR over a year ago, I'm realizing as I write this up now 🙈

For example:

```js
"aud": "did:key:z6MkiTBz1ymuepAQ4HEHYSF1H8quG5GLVVQR3djdX3mDooWp",
"iss": "did:key:zDnaerDaTF5BXEavCrfRZEk316dpbLsfPDZ3WJ5hRTPFU2169",
```

> Please see the [Cryptosuite] section for more detail on DIDs.

Reviewer 2: Narrator: but there was no more detail on DIDs there.

Reviewer 2: ```suggestion
Please see the [Cryptosuite](#cryptosuite) section for more detail on DIDs.
```

Below are a couple examples:

```js
{
  // ...
  "nbf": 1529496683,
  "exp": 1575606941
}
```

```js
{
  // ...
  "exp": 1575606941
}
```

```js
{
  // ...
  "nbf": 1529496683,
  "exp": null
}
```

## Lifecycle Example

Here is a concrete example of all stages of the UCAN lifecycle for database write access.

``` mermaid
sequenceDiagram
    participant Database

    actor DBAgent
    actor Alice
    actor Bob

    Note over Database, DBAgent: Set Up Agent-Owned Resource
    DBAgent ->> Database: createDB()

    autonumber 1

    Note over DBAgent, Bob: Delegation
    DBAgent -->> Alice: delegate(DBAgent, write)
    Alice -->> Bob: delegate(DBAgent, write)

    Note over Database, Bob: Invocation
    Bob ->> DBAgent: invoke(DBAgent, [write, [key, value]], proof: [➊,➋])
    DBAgent ->> Database: write(key, value)
    DBAgent ->> Bob: ACK

    Note over DBAgent, Bob: Revocation
    Alice ->> DBAgent: revoke(➋, proof: [➊,➋])
    Bob ->> DBAgent: invoke(DBAgent, [write, [key, newValue]], proof: [➊,➋])
    DBAgent -X Bob: NAK(➏) [rejected]
```

## Capability

A capability is the association of an ability to a subject: `subject x command x policy`.

> The Subject and Command fields are REQUIRED. Any non-normative extensions are OPTIONAL.

Reviewer 2: are all policies "non-normative extensions"? are any other fields?  OPTIONAL for implementations to handle?  Dropped if unknown?  I know this is all stated in the delegation spec, but just speaking as a reader, a tiny bit more detail, although redundant, might be helpful here for high-level readers!

Reviewer 2: also, are they "Arguments"  now? I am seeing links elsewhere in this document to `[Arguments]` but there is no `### Arguments` section (or for that matter a `### Policy` or `### Extensions` section, either!), so probably time to add one.

For example, a capability may used to represent the ability to send email from a certain address to others at `@example.com`.

| Field      | Example                                                                                      |
|------------|----------------------------------------------------------------------------------------------|
| Subject    | `did:key:z6MkhaXgBZDvotDkL5257faiztiGiC2QtKLGpbnnEGta2doK`                                   |

> | Command    | `msg/send`                                                                                   |

Reviewer 3: Since we have moved to paths in delegation

```suggestion
| Command    | `/msg/send`                                                                                  |
```

The wildcard ability grants access to all other capabilities for the specified resource, across all possible namespaces. The wildcard ability is useful when "linking" agents by delegating all access to another device controlled by the same user, and that should behave as the same agent. It is extremely powerful, and should be used with care. Among other things, it permits the delegate to update a Subject's mutable DID document (change their private keys), revoke UCAN delegations, and use any resources delegated to the Subject by others.

``` mermaid
%%{ init: { 'flowchart': { 'curve': 'linear' } } }%%

flowchart BT
  /

  /msg --> /
  subgraph msgGraph [ ]
    /msg/send --> /msg
    /msg/receive --> /msg
  end

  /crud --> /
  subgraph crudGraph [ ]
    /crud/read --> /crud
    /crud/mutate --> /crud

    subgraph mutationGraph [ ]
        /crud/mutate/create --> /crud/mutate
        /crud/mutate/update --> /crud/mutate
        /crud/mutate/destroy --> /crud/mutate
    end
  end

  ... --> /
```

### Reserved Commands

#### `/ucan` Namespace

> The `/ucan` Command namespace MUST be reserved. This MUST include any ability string matching the regex `^ucan\/.*`. This is important for keeping a space for community-blessed Commands in the future, such as standard library Commands, such as [Revocation].

Reviewer 2: ```suggestion
The `/ucan` Command namespace MUST be reserved. This MUST include any ability string matching the regex `^ucan(\/(^\m\/)+)+`. This is important for keeping a space for community-blessed Commands in the future, such as standard library Commands, such as [Revocation].
```
I'm no regex wizard but I'm guessing this regex is older than the prohibition on trailing `/`s, which this newer, uglier regex bans.

Also, I couldn't tell if `/ucan` was supposed to be a valid command (or a valid blank-check for future stdlib commands), but if so, that last `+` (1 or more) should be a `*` (0 or more).

## Attenuation

> Attenuation is the process of constraining the capabilities in a delegation chain. Each direct delegation MUST either directly restate or attenuate (diminish) its capabilities.

Reviewer 2: ```suggestion
Attenuation is the process of constraining the capabilities in a delegation chain. Each direct delegation MUST either explicitly restate or attenuate (diminish) its capabilities.
```

# Token Resolution

Token resolution is transport specific. The exact format is left to the relevant UCAN transport specification. At minimum, such a specification MUST define at least the following:

1. Request protocol
2. Response protocol
3. Collections format

> Note that if an instance cannot dereference a CID at runtime, the UCAN MUST fail validation. This is consistent with the [constructive semantics] of UCAN.

Reviewer 2: Note that this is the first mention of a CID in the entire high-level spec, without a normref or definition.

Editorial nitpicking aside, I think it might be clearer to refer to them as `link`s in this and other section (which, in IPLD, are effectively just IPFS CIDs) rather than refer to them as CIDs, because what matters is that they are identifiers they can't be defererenced, not anything about their nature as _content_ identifiers or multiformats, etc etc.  (I.e., I can imagine an RDF/JSON-LD version of this where the same link-dereferenceability rule applies to DID URLs, or HTTPS URLS, or some other kinds of dereferenceable URIs...)

Reviewer 2: For example:
```suggestion
Note that if a UCAN includes links in the form of [IPFS CIDs][CID] to subgraphs, and a parsing implementation cannot expand the full graph because one or more of these links cannot be dereferenced at runtime, then the UCAN MUST fail validation. This is consistent with the [constructive semantics] of UCAN.
```

# Nonce

The REQUIRED nonce parameter `nonce` MAY be any value. A randomly generated string is RECOMMENDED to provide a unique UCAN, though it MAY also be a monotonically increasing count of the number of links in the hash chain. This field helps prevent replay attacks and ensures a unique CID per delegation. The `iss`, `aud`, and `exp` fields together will often ensure that UCANs are unique, but adding the nonce ensures uniqueness.

The recommended size of the nonce differs by key type. In many cases, a random 12-byte nonce is sufficient. If uncertain, check the nonce in your DID's crypto suite.

> This field SHOULD NOT be used to sign arbitrary data, such as signature challenges. See the [`meta`][Metadata] field for more.

Reviewer 2: ```suggestion
This field SHOULD NOT be used to sign arbitrary data, such as signature challenges. See the [`meta`][#Metadata] field for more.
```

Here is a simple example.

``` js
{
  // ...
  "nonce": {"/": {"bytes": "bGlnaHQgd29yay4"}}

> }

Author: > Reading quickly I got thrown off by the "/" and thought it was an IPLD thing.

it is an IPLD thing. This is how DAG-JSON represents binary data

```

# Metadata

> The OPTIONAL `meta` field contains a map of arbitrary metadata, facts, and proofs of knowledge. The enclosed data MUST be self-evident and externally verifiable. It MAY include information such as hash preimages, server challenges, a Merkle proof, dictionary data, etc.

Reviewer 2: ```suggestion
The OPTIONAL `meta` field contains a map of arbitrary metadata, facts, and proofs of knowledge. The enclosed data SHOULD be self-evident and externally verifiable. It MAY include information such as hash preimages, server challenges, a Merkle proof, dictionary data, etc.
```
It feels like this MUST is untestable, and is more like an instruction to UCAN-generating/encoding implementations than anything that could be included in a conformance test? It might be better phrased as a consuming behavior, i.e., "Validators MUST NOT consider the contents when proofing UCAN invocation".

corner-case: do CIDs in metadata have to be dereferenceable at runtime as well? or does their being unmeaningful trump their being IPLD 😄

The data contained in this map MUST NOT be semantically meaningful to delegation chains.

Below is an example:

``` js
{
  // ...
  "meta": {
    "challenges": {
      "example.com": "abcdef",
      "another.example.net": "12345"
    },

>     "sha3_256": {

Reviewer 2: not `SHA-256`, i.e. SHA2? or is that deliberate?

Author: This is an example of putting in arbitrary data. It literally doesn't matter

      "B94D27B9934D3E08A52E52D7DA7DABFAC484EFE37A5380EE9088F7ACE2EFCDE9": "hello world"
    }
  }
}
```

# Canonicalization

## Cryptosuite

Across all UCAN specifications, the following cryptosuite MUST be supported:

| Role      | REQUIRED Algorithms               | Notes                                | 
|-----------|-----------------------------------|--------------------------------------|
| Hash      | [SHA-256]                         |                                      |
| Signature | [Ed25519], [P-256], [`secp256k1`] | Preference of Ed25519 is RECOMMENDED |
| [DID]     | [`did:key`]                       |                                      |

## Encoding

> All UCANs MUST be canonically encoded with [DAG-CBOR] for signing. A UCAN MAY be presented or stored in other [IPLD] formats (such as [DAG-JSON]), but converted to DAG-CBOR for signature validation.

Reviewer 2: I feel like the ` { "/" : { "bytes": "...` notation needs to be explained a little here, rather than assuming the reader both wants to and can reasonably click that "DAG-CBOR" link and come back 10 minutes later knowing what IPLD is, if they don't already... because if _I_ am confused, who is your target reader 😏

## Content Identifiers

A UCAN token MUST be configured as follows:

| Parameter    | REQUIRED Configuration |
|--------------|------------------------|
| Version      | [CIDv1]                | 
| [Multibase]  | [`base58btc`]          | 
| [Multihash]  | [SHA-256]              |
| [Multicodec] | [DAG-CBOR]             |

> [!NOTE]

> > All CIDs encoded as above start with the characters `zdpu`.

Reviewer 2: If they're canonically DAG-CBOR why is the multibase needed, as opposed to [staying in binary](https://github.com/multiformats/cid?tab=readme-ov-file#how-does-it-work)? Maybe I've been talking to the wrong people but I thought CIDv1 was _canonically_ binary and _optionally_ multibased, so it feels like an extra roundtrip to multibase it and then drop back down to CBOR...

> The resolution of these addresses is left to the implementation and end-user, and MAY (non-exclusively) include the following: local store, a distributed hash table (DHT), gossip network, or RESTful service.

Reviewer 2: Redundant, sure, but here might be a good place to remind the reader that a validator MUST fail verification of a UCAN with subgraphs linked by CID that do not dereference at time of consumption of invocations.

    "ucan/example@1.0.0-rc.1": {
      "hello": "world"
    }
  }
]
```

### Payload

A UCAN's Payload MUST contain at least the following fields:

| Field   | Type                                      | Required | Description                                                 |
|---------|-------------------------------------------|----------|-------------------------------------------------------------|
| `iss`   | `DID`                                     | Yes      | Issuer DID (sender)                                         |
| `aud`   | `DID`                                     | Yes      | Audience DID (receiver)                                     |
| `sub`   | `DID`                                     | Yes      | Principal that the chain is about (the [Subject])           |
| `cmd`   | `String`                                  | Yes      | The [Command] to eventually invoke                          |
| `args`  | `{String : Any}`                          | Yes      | Any [Arguments] that MUST be present in the Invocation      |
| `nonce` | `Bytes`                                   | Yes      | Nonce                                                       |

> | `meta`  | `{String : Any}`                          | No       | [Meta] (asserted, signed data) — is not delegated authority |

Reviewer 2: Maybe break out the non-required fields into a separate table below, since all of this is under MUST CONTAIN?

And throw `iat` in there, too?

Author: The rationale is that `iat` is not in the delegation spec since it's not particularly helpful for delegation. Invocation does have an `iat`, so it's not a common field across all UCAN formats

| `nbf`   | `Integer` (53-bits[^js-num-size])         | No       | "Not before" UTC Unix Timestamp in seconds (valid from)     |
| `exp`   | `Integer \| Null` (53-bits[^js-num-size]) | Yes      | Expiration UTC Unix Timestamp in seconds (valid until)      |

# Implementation Recommendations

## Delegation Store

A validator MAY keep a local store of UCANs that it has received. UCANs are immutable but also time-bound so that this store MAY evict expired or revoked UCANs.

This store SHOULD be indexed by CID (content addressing). Multiple indices built on top of this store MAY be used to improve capability search or selection performance.

## Memoized Validation

> Aside from revocation, capability validation is idempotent. Marking a CID (or capability index inside that CID) as valid acts as memoization, obviating the need to check the entire structure on every validation. This extends to distinct UCANs that share a proof: if the proof was previously reviewed and is not revoked, it is RECOMMENDED to consider it valid immediately.

Reviewer 2: Should that by "Aside from revocation **and expiration**,..." ? I am also confused by the corner case mentioned here by Gozala: 

Let's say Alice delegates to Bob a capability `/crud/` over a hard disk, then Bob sends Carol (me) a capability `/crud/write/`.  Before I go to invoke it at Alice's hard disk, Alice's original parent cap expired or was revoked.  Bob sends me a new sub-delegation identical except for having a fresher parent.  That's a different token, right? This idempotency stuff doesn't apply because if it's a different chain, it's a different token?

> Revocation is irreversible. Suppose the validator learns of revocation by UCAN CID. In that case, the UCAN and all of its derivatives in such a cache MUST be marked as invalid, and all validations immediately fail without needing to walk the entire structure.

Reviewer 2: I'd love for "Suppose the validator learns of revocation by UCAN CID" to be expanded a bit, I'm having trouble supposing from that very high-level summary 😅

[Seitan token exchange]: https://book.keybase.io/docs/teams/seitan
[Steven Vandevelde]: https://github.com/icidasset
[Storacha]: https://storacha.network/
[The Structure of Authority]: http://erights.org/talks/no-sep/secnotsep.pdf
[The computer revolution hasn't happened yet]: https://www.youtube.com/watch?v=oKg1hTOQXoY
[UCAN Promise]: https://github.com/ucan-wg/promise
[URI]: https://www.rfc-editor.org/rfc/rfc3986
[Varsig]: https://github.com/ChainAgnostic/varsig
[Verifiable credentials]: https://www.w3.org/2017/vc/WG/
[W3C]: https://www.w3.org/
[WebCrypto API]: https://developer.mozilla.org/en-US/docs/Web/API/Web_Crypto_API
[Witchcraft Software]: https://github.com/expede
[ZCAP-LD]: https://w3c-ccg.github.io/zcap-spec/
[`base58btc`]: https://github.com/multiformats/multibase/blob/master/multibase.csv#L21
[`did:key`]: https://w3c-ccg.github.io/did-method-key/
[`secp256k1`]: https://en.bitcoin.it/wiki/Secp256k1
[browser api crypto key]: https://developer.mozilla.org/en-US/docs/Web/API/CryptoKey
[capabilities]: https://en.wikipedia.org/wiki/Object-capability_model
[caps as keys]: http://www.erights.org/elib/capability/duals/myths.html#caps-as-keys

> [certificate capability model]: http://wiki.erights.org/wiki/Capability-based_Active_Invocation_Certificates

Reviewer 2: ```suggestion
[certificate capability model]: http://wiki.erights.org/wiki/Capability-based_Active_Invocation_Certificates
[CID]: https://docs.ipfs.tech/concepts/content-addressing/
```
Unless the uglier, raw github definition seems more appropriate or canonical for whatever reason? 
https://github.com/multiformats/cid
